t/op/signatures.t

```diff
@@ -1608,6 +1610,49 @@ while(<$kh>) {
             'f($1)';
 }
 
+# check that various uses of @_ inside signatured subs causes "discouraged"
+# warnings at compiletime
+{
+    sub discouraged_ok {
+        my ($opname, $code) = @_;
+        my $warnings = "";
+        local $SIG{__WARN__} = sub { $warnings .= join "", @_ };
+        eval qq{ sub(\$x) { $code }};
+        ok($warnings =~ m/[Uu]se of \@_ in $opname is discouraged in signatured subroutine at \(eval /,
+            "`$code` warns of discouraged \@_") or
+            diag("Warnings were:\n$warnings");
+    }
+
+    # implicit @_
+    discouraged_ok 'shift',            'shift';
+    discouraged_ok 'pop',              'pop';
+    discouraged_ok 'subroutine entry', 'goto &SUB'; # tail-call
+    discouraged_ok 'subroutine entry', '&SUB'; # perl4-style
+
+    # explicit @_
+    discouraged_ok 'shift',            'shift @_';
+    discouraged_ok 'pop',              'pop @_';
+    discouraged_ok 'array element',    '$_[0]';
+    discouraged_ok 'array element',    'my $one = 1; $_[$one]';
+    discouraged_ok 'push',             'push @_, 1';
+    discouraged_ok 'unshift',          'unshift @_, 9';
+    discouraged_ok 'splice',           'splice @_, 1, 2, 3';
+    discouraged_ok 'keys on array',    'keys @_';
+    discouraged_ok 'values on array',  'values @_';
+    discouraged_ok 'each on array',    'each @_';
+    discouraged_ok 'print',            'print "a", @_, "z"';
+    discouraged_ok 'subroutine entry', 'func("a", @_, "z")';
```

Comment:
an extra test for `eval`?
` discouraged_ok 'eval',    'eval qq[ $_[0] ]';`

Reply:
The `eval` one is potentially interesting. Not sure about the `for on array` though, because that's really just the same as most of the others. The implementation happens at a lower level than that so if any warnings work they all would.

Not sure what you're aiming at with those `__ANON__` ones.

Reply:
yeap the `__ANON__` has probably low value and is a dupe of any other subroutine call

Reply:
thanks for adding the `eval` one